Feature/review flow

MatejMa2ur opened the pull request:
- Closes: #25 
- Closes: #26 
- Closes: #27 
- Closes: #28

MatejMa2ur committed:
Add IsDeleted property and ReviewMapper for review handling

Introduced the `IsDeleted` property in Review entities to support soft deletion. Added `ReviewMapper` to handle conversions between entities, DTOs, and commands. Enhanced `CreateReviewCommand` and `ReviewGetDto` to include new fields for comprehensive review management.

MatejMa2ur committed:
Implement review CRUD operations for read and write repositories

Added async methods to manage reviews in both PostgreSQL and MongoDB, including Create, Read, Update, and Delete operations. Integrated event handlers to synchronize write and read models using domain events for consistency.

MatejMa2ur committed:
Implement review CRUD operations with validation and events

Added Create, Update, and Delete review commands with validation logic and proper exception handling. Integrated domain event publishing and mapped results to DTOs. Updated command handler and service implementations to ensure functionality.

MatejMa2ur committed:
Set ReviewId in DeleteReviewAsync command

This change ensures the ReviewId is correctly passed to the DeleteReviewCommand in DeleteReviewAsync. Without this, the deletion process may fail due to missing data.

juuwel commented on Reclaim/Application/Commands/Review/UpdateReviewCommand.cs:
```diff
@@ -1,7 +1,14 @@
 public class UpdateReviewCommand : ICommand<ReviewWriteEntity>
 {
+    [MaxLength(24)]
+    public required string ReviewId { get; set; } = string.Empty;
```
Is there a specific reason we are initializing the property despite the fact that it is required?

```suggestion
    public required string ReviewId { get; set; }
```

juuwel commented on Reclaim/Domain/DTOs/ReviewGetDto.cs:
Should we not make the ID and CreatedAt required to keep it consistent with the entities? I see in the mapper that you are mapping these properties either way, and I don't believe we will ever run into a situation where we are constructing a DTO without these two properties (maybe without created at, but definitely not without an ID)

MatejMa2ur replied:
That's why I am using the `null!`

I recommend to look into difference between the required and null! : https://chwastek.eu/blog/handling-nullability-and-required-properties-in-dotnet-api-models#:~:text=TL%3BDR%20You%20can%20use%20either%20the%20required%20keyword,without%20required%20property%2C%20you%20get%20model%20binding%20error.
They kinda do the same thing, but not really, I am used for DTOs mostly use the `null!`

juuwel replied:
No need to change it, let's keep it as is, just wanted to finish my thought.

It does sound like it is a different approach when you need to handle a request's validation, but the GetDTO is something always returned by the server, not supplied by the client right? So in that case, required would maybe be a bit more valuable than null!, since with null!, we would not be reminded by the compiler to set the property when we are making an instance of the GetDTO. But like I said, no need to change it, just wanted to make sure I understand that right.

juuwel commented on Reclaim/Infrastructure/Repositories/Write/Implementations/ReviewWriteEfRepository.cs:
```diff
@@ -1,33 +1,63 @@
+    public async Task<IEnumerable<ReviewWriteEntity>> GetAllAsync(bool includeDeleted = false)
     {
-        throw new NotImplementedException();
+        var query = dbContext.Reviews
+            .Include(r => r.Author)
+            .Include(r => r.Seller)
+            .Where(r => includeDeleted || !r.IsDeleted)
+            .AsQueryable();
+        
+        return await query.ToListAsync();
```
This is just for my sake, is there a benefit to using AsQueryable() and then calling ToListAsync() on that instead of calling it directly?

MatejMa2ur replied:
Nope not now, I was experimenting a little with the where.
I understood it a little differently.

juuwel reviewed (approved):
Just a few small comments, but otherwise looks great 🚀

MatejMa2ur commented:
>[!CAUTION]
> Do not merge yet!

MatejMa2ur committed:
Update review query and repository implementations

Added required properties with validation for query DTOs and updated the query handler to fetch reviews from the repository. Adjusted repository methods to return lists instead of single entities, aligning with the query requirements. Simplified query building in repositories and ensured proper mapping to DTOs where necessary.
